### PR TITLE
Remove bash dependency from init-compiler

### PR DESCRIPTION
**Pull request by am11**
Also fixes shell check warnings in `System.Globalization.Native/local_build.sh`.

Upstream PR: https://github.com/dotnet/arcade/pull/11359

### CODE REVIEW

**Comment by am11** [comment]
cc @jkotas, @akoeplinger
This is in preparation for including `init-compiler.sh` to NativeAOT package: https://github.com/dotnet/runtime/pull/77245#issuecomment-1285330856. Some OS (e.g. Alpine Linux) don't have bash preinstalled, and we aren't using many bash features besides two usages of bash arrays, so it's simple to replace them and drop the dependency from this standalone script.

**Comment by jkotas** [comment]
> This is in preparation for including init-compiler.sh to NativeAOT package

I am not sure whether it is a good idea to include `init-compiler.sh` in the NativeAOT package. I see this script as specific to the .NET runtime internal build infrastructure setup. It is why it lives in Arcade repo. As a rule, bits that live in Arcade repo should be only used by .NET runtime internal build infrastructure. They should not ship as part of the product.

The current native AOT plan is: You have to have version-less tools installed. If you do not have version-less tools installed, you have to set `CppCompilerAndLinker` to point to the tools that you want to use.

**Comment by am11** [comment]
> As a rule, bits that live in Arcade repo should be only used by .NET runtime internal build infrastructure.

We moved this script to arcade to avoid duplicating the code: https://github.com/dotnet/runtime/pull/59018 (it is already included in `Microsoft.DotNet.CMake.Sdk` nuget package).

This is a freestanding script to locate compiler on the system. It works for all supported build environments (it does not look for prerequisite beyond a few compiler toolchain components, mainly the driver; clang or gcc).

> you have to set `CppCompilerAndLinker` to point to the tools that you want to use.

I was experimenting (for a follow up PR) with keeping that experience. If user specified e.g. `CppCompilerAndLinker=clang-15` or there is `clang` or `gcc` found in `PATH`, then we will not invoke the script. Without setting `CppCompilerAndLinker` or when there is no `clang` or `gcc` in PATH, we will invoke this script to find the best available versioned compiler found in `PATH` (descending order of their versions).

**Comment by jkotas** [comment]
> it is already included in Microsoft.DotNet.CMake.Sdk

This nuget package is not a shipping nuget package. It is not on nuget.org. It is for internal .NET build infrastructure use only.

**Comment by am11** [comment]
> This nuget package is not a shipping nuget package. It is not on nuget.org. It is for internal .NET build infrastructure use only.

Right, my point was the only reason it's moved to arcade was to avoid maintaining two copies of same script, that locates versioned compiler in PATH.

**Comment by akoeplinger** [comment]
It would still mean we're essentially signing up for shipping init-compiler.sh officially (even if it's just for this opt-in static ICU linking feature from local_build.sh).

That said, maybe that's the right choice given we need to hardcode clang version numbers which are specific to our build environment e.g. like this today: https://github.com/dotnet/runtime/blob/f4b715f1cfadcbf85486f369724528c051be5733/src/coreclr/tools/aot/crossgen2/crossgen2.csproj#L34-L35

**Comment by jkotas** [comment]
> That said, maybe that's the right choice given we need to hardcode clang version numbers which are specific to our build environment e.g. like this today:

We can call ` init-compiler.sh` from our build environment to fix these hardcoded version numbers. It is not necessary to ship `init-compiler.sh` for that.

**Comment by am11** [comment]
Removed hard-coded versions in favor of similar target used in mono and DNNE test projs.

File: src/coreclr/tools/aot/crossgen2/crossgen2.csproj
Comment (akoeplinger) [inline src/coreclr/tools/aot/crossgen2/crossgen2.csproj:67]: looks like this is resolving gcc now in the `runtime (Build Installer Build and Test coreclr Linux_arm64 Release` build:

```
  Generating native code
gcc : error : unrecognized command line option ‘--target=aarch64-linux-gnu’ [/__w/1/s/src/coreclr/tools/aot/crossgen2/crossgen2.csproj]

```

Comment (am11) [reply src/coreclr/tools/aot/crossgen2/crossgen2.csproj:67]: This is only failing after merge from `main` because property name was changed there: https://github.com/dotnet/runtime/commit/8e15d9f523209caf08db57897777b7beb0516c99

File: src/coreclr/tools/aot/crossgen2/crossgen2.csproj
Comment (am11) [inline src/coreclr/tools/aot/crossgen2/crossgen2.csproj:61]:
```suggestion
          Condition="'$(_IsPublishing)' == 'true' and '$(HostOS)' != 'windows'"
```

File: eng/testing/tests.singlefile.targets
Comment (akoeplinger) [inline eng/testing/tests.singlefile.targets:86]: this should probably use HostOS, not TargetOS

or should we keep the `'$(TargetArchitecture)' != '$(BuildArchitecture)' and '$(HostOS)' == 'linux'` that was there initially?

File: eng/testing/tests.singlefile.targets
Comment (akoeplinger) [inline eng/testing/tests.singlefile.targets:96]: I assume this should be `CppCompilerAndLinker` ?

File: eng/testing/tests.singlefile.targets
Comment (am11) [reply eng/testing/tests.singlefile.targets:86]: > this should probably use HostOS, not TargetOS

For Windows it doesn't make a difference as we don't support Windows -> Linux or Linux -> Windows cross compilation.

> or should we keep the `'$(TargetArchitecture)' != '$(BuildArchitecture)' and '$(HostOS)' == 'linux'` that was there initially?

That would mean we we only want to run the init-compiler script from this target for cross compilation on Linux, but I wanted to always run this target on all Unix-like OS as we do it for rest of the native components so we can uniformly resolve the exact compiler which we used elsewhere. This will fix issues like https://github.com/dotnet/runtime/pull/77245#issuecomment-1285330856 in non-CI / external-CI use-cases.

Comment (akoeplinger) [reply eng/testing/tests.singlefile.targets:86]: ok. I'd still prefer $(HostOS) since it makes more sense and it's also what we're using in crossgen2.csproj

File: src/coreclr/tools/aot/crossgen2/crossgen2.csproj
Comment (akoeplinger) [inline src/coreclr/tools/aot/crossgen2/crossgen2.csproj:61]: I think we should also check for `'$(NativeAotSupported)' == 'true'`

File: eng/testing/tests.singlefile.targets
Comment (akoeplinger) [reply eng/testing/tests.singlefile.targets:96]: ok makes sense, would you mind adding a short comment since it was not immediately obvious to me that this overrides what gets set in the nativeaot targets

File: src/coreclr/tools/aot/crossgen2/crossgen2.csproj
Comment (am11) [reply src/coreclr/tools/aot/crossgen2/crossgen2.csproj:61]: Yes, thanks!
```suggestion
          Condition="'$(NativeAotSupported)' == 'true' and '$(_IsPublishing)' == 'true' and '$(HostOS)' != 'windows'"
```

File: eng/testing/tests.singlefile.targets
Comment (am11) [inline eng/testing/tests.singlefile.targets:85]:
```suggestion
  <!--
    Use init-compiler.sh to locate the compiler toolchain which was resolved for rest of the build.
    This target is essentially an override hook which is called before `SetupOSSpecificProps` in
    `Microsoft.NETCore.Native.Unix.targets`. Note that the input is `CppCompilerAndLinker`
    and the output is `CppLinker`, becuase from `SetupOSSpecificProps` onwards, we only use `CppLinker`
    (when alternative compiler, i.e. gcc, is not selected)
  -->
  <Target Name="LocateNativeCompiler"
```

Comment (jkotas) [reply eng/testing/tests.singlefile.targets:85]: becuase -> because

Comment (akoeplinger) [reply eng/testing/tests.singlefile.targets:85]: Excellent, thanks

File: eng/testing/tests.singlefile.targets
Comment (akoeplinger) [inline eng/testing/tests.singlefile.targets:96]: looking at the whole PR again, maybe it'd be better to not use this property name at all and go with what we use in the NativeExports.csproj:
```
    <PropertyGroup>
      <NativeCompiler>$(Compiler)</NativeCompiler>
      <NativeCompiler Condition="'$(NativeCompiler)' == ''">clang</NativeCompiler>
    </PropertyGroup>
```

that'd also support overriding via the `$(Compiler)` property.

Comment (am11) [reply eng/testing/tests.singlefile.targets:96]: Compiler == CppCompilerAndLinker here, see eng/build.sh

Comment (akoeplinger) [reply eng/testing/tests.singlefile.targets:96]: omg, so many abstractions 😄

**Review by akoeplinger — APPROVED** [review]

File: eng/testing/tests.singlefile.targets
Comment (am11) [reply eng/testing/tests.singlefile.targets:96]: :)

I am all for replacing `CppCompilerAndLinker` and `CppLinker` in favor of `NativeCompiler` as property name, but I think some consumers are relying on `CppCompilerAndLinker`.

File: src/native/libs/System.Globalization.Native/local_build.sh
Comment (am11) [inline src/native/libs/System.Globalization.Native/local_build.sh:19]:
```suggestion
    if ! { mkdir -p "$LOCAL_SHIM_DIR" && cd "$LOCAL_SHIM_DIR"; } 2>/dev/null; then
```

File: src/native/libs/System.Globalization.Native/local_build.sh
Comment (am11) [inline src/native/libs/System.Globalization.Native/local_build.sh:19]:
```suggestion
    if ! { mkdir -p "$LOCAL_SHIM_DIR" && cd "$LOCAL_SHIM_DIR"; }; then
```

**Comment by am11** [comment]
Failures are unrelated and known: https://github.com/dotnet/runtime/issues/79843 (mono llvmfullaot), https://github.com/dotnet/runtime/issues/79439 (win-x64) and https://github.com/dotnet/runtime/issues/79569 (browser-wasm).

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**
